library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -304,7 +304,29 @@ final class RuntimeLong(val lo: Int, val hi: Int)
   def *(b: RuntimeLong): RuntimeLong = {
     val alo = a.lo
     val blo = b.lo
-    new RuntimeLong(alo * blo, RuntimeLong.timesHi(alo, a.hi, blo, b.hi))
+
+    val a0 = alo & 0xffff
+    val a1 = alo >>> 16
+    val b0 = blo & 0xffff
+    val b1 = blo >>> 16
+
+    val a0b0 = a0 * b0
+    val a1b0 = a1 * b0 // collapses to 0 when a is constant and 0 <= a <= 0xffff
+    val a0b1 = a0 * b1
+
+    /* lo = a.lo * b.lo, but we compute the above 3 subproducts for hi
+     * anyway, we reuse them to compute lo too, trading a * for 2 +'s and 1 <<.
+     */
+    val lo = a0b0 + ((a1b0 + a0b1) << 16)
+
+    // hi = a.lo*b.hi + a.hi*b.hi + carry_from_lo_*
+    val c1part = (a0b0 >>> 16) + a0b1
+    val hi = {
+      alo*b.hi + a.hi*blo + a1 * b1 + (c1part >>> 16) +
```

Comment:
No, it is not (yet) smart enough for that. But as explained above, `a0b1 == 0` is not a likely scenario.